# Модуль 3. Математические операции с массивами данных

# Выполнение операций над массивами и таблицами с помощью модуля `Numpy`

В ходе прошлых занятий вы работали с последовательностями – типами данных, позволяющими хранить множество различных значений произвольных типов. При численном решении физических задач вам чаще всего придётся использовать списки, поскольку они являются изменяемым типом данных, позволяющим не только хранить упорядоченные последовательности элементов, но и удобно оперировать ими.

Наряду с перечисленными преимуществами, тип данных `list` обладает и существенным ограничением – низкой скоростью обработки данных, сужающим возможности его применения при решении задач, требующих работы с большими объёмами информации. Это ограничение связано с тем, что элементами списка могут быть различные типы данных, так что интерпретатору при обращении к каждому элементу списка требуется дополнительное время, в т. ч. для проверки его типа.

Для решения этой проблемы в `Python` была разработана библиотека (или модуль, если говорить в терминах `Python`) **`NumPy`**. Ниже мы рассмотрим основной функционал модуля `NumPy`, который может быть использован при решении задач.

## Массивы `NumPy`. Объявление массивов, заполнение с использованием цикла. Заполнение случайными числами. Извлечение из массива элемента с заданным номером.

Пакет `NumPy` определяет новый тип данных – $N$-мерный массив (`ndarray`). Массив – это упорядоченный набор элементов для хранения данных *одного* типа, идентифицируемых с помощью одного или нескольких индексов. 

Благодаря тому, что тип элементов массива `NumPy` един, интерпретатору нет необходимости проверять его каждый раз при обращении к тому или иному элементу, что значительно сокращает время на обработку данных.

### Объявление массивов

Для создания массива необходимо подключить модуль `numpy` (так, как это показано на примере ниже) и вызвать функцию `array`. Функция принимает в качестве аргумента любую последовательность (список, кортеж и т. д.) и преобразует его в массив `ndarray`. При этом важно следить за тем, чтобы все элементы последовательности имели один и тот же тип!

In [1]:
import math  # подключаем модуль Math – для работы с математическими функциями
import numpy as np  # "np" – это псевдоним, сокращенное имя модуля numpy

# преобразуем список списков в массив ndarray
a = np.array([[1, math.e, math.pi], [4, 5, 6]])

print(a, end='\n\n')  # выводим на печать элементы массива,
print(a.shape)        # его размерность
print(a.dtype)        # и тип элементов
print()

# попробуем сделать массив-компот из данных разных типов
compote = np.array([[1e0, 'e', 'Pi'], [(4), 'five', 6]])
print(compote)
print(compote.dtype)
print(type(compote))
print(type(compote[1, 0]))


[[1.         2.71828183 3.14159265]
 [4.         5.         6.        ]]

(2, 3)
float64

[['1.0' 'e' 'Pi']
 ['4' 'five' '6']]
<U32
<class 'numpy.ndarray'>
<class 'numpy.str_'>


Как видите, массив-компот все же удалось создать, но все его элементы превратились в строки (метод `dtype` выдал "U32", что означает строковые данные в *unicode*).

Обратите внимание на то, что созданный массив `a` обладает рядом атрибутов.

* Атрибут `shape` (буквально, "форма") – кортеж, в котором хранятся размерности массива, – число элементов вдоль каждой оси массива (строки и столбцы в соответствующем порядке).
* Атрибут `dtype` (сокращенно от data type) – возвращает тип элементов, которые хранятся в массиве.

Общая рекомендация при работе с массивами: лучше всегда в явном виде указывать тип данных, которые вы хотите поместить в этот массив. Это поможет избежать трудно отслеживаемых опечаток, приводящих к преобразованию данных и потере точности. Таким образом, если бы мы записали массив `compote` следующим образом:
~~~python
compote = np.array([[1e0, 'e', 'Pi'], [(4), 'five', 6]], dtype=float)
~~~
то интерпретатор выдал бы ошибку.

Полный список атрибутов, как и в рассмотренных ранее примерах, можно получить, вызвав команду:
~~~python
dir(a)
~~~
или
~~~python
help(np)  # работает после импортирования модуля
~~~

Также отметим то, как массив выводится на печать: внешне он выглядит как "список списков": внешняя пара квадратных скобок ограничивает весь массив, а каждая строка в нем заключена в свою пару квадратных скобок. При этом элементы в каждой строке выстроены так, чтобы визуально отделить столбцы (в вывод включены символы табуляции `'\t'`).

##### Два способа задать массив
Функция `np.array()` позволяет создавать массив из уже существующей последовательности (иными словами, уже **заполненный данными**). Для создания массива, элементы которого не инициализированы в начальный момент (**пустой массив**), необходимо использовать функцию `np.ndarray()`. Обязательный аргумент этой функции – размерность массива – задается кортежом значений.

Рассмотрим пример:

In [2]:
import numpy as np  # подключаем NumPy

a = np.ndarray((3, 4, 5))  # создаём трехмерный массив: 3 x 4 x 5 элементов
b = np.ndarray(a.shape, dtype = float)  # создаём массив по форме массива "a"

# выводим содержимое массивов на печать
print(a)
print('\n' + '-'*70 + '\n')  # в качестве визуального разделителя между массивами
print(b)


[[[ 8.15979057e-312  8.15979075e-312  9.73469815e-309 -5.48612429e+303
    1.11335178e-308]
  [ 3.75481214e-308  2.48104029e-265  7.29112297e-304  8.70018274e-313
    1.71881659e-312]
  [ 3.41641322e-312  1.48539706e-313  1.45042656e-309  3.72366666e+150
    7.37247701e+228]
  [ 2.78694543e+199  9.44905842e+170  1.90288078e+045  6.01184167e-067
    1.90762878e+011]]

 [[ 9.56706096e+005  1.91621338e-076  1.72717729e-042  9.77814434e+199
    1.35874057e-312]
  [ 0.00000000e+000  5.43230923e-311  6.05220090e-306  1.40918088e-293
    3.13151381e-294]
  [ 2.36279258e-313  7.35876461e+223  9.22950641e+179  3.64253113e+185
    1.63732307e-306]
  [ 5.56270587e-309  6.95335581e-308  4.61390353e-270  3.13151575e-294
    2.33626763e-313]]

 [[ 4.25898788e-270  2.10541771e-314  5.61870786e-310  6.10949039e+093
    7.49654801e-304]
  [ 2.17173007e-314  2.05835965e-312  2.36414097e-308  5.56268466e-309
    2.84809528e-306]
  [ 4.36671189e+242  1.64080109e-306  3.57225739e-057  3.03387206e-152
    1

Обратите внимание, что созданные при помощи команды `np.ndarray()` массивы можно вывести на печать, однако в их элементах хранится "абракадабра" – случайный набор значений, величина которых близка к нулю или к бесконечности (а в массиве `a` даже встречается значение `nan` – *not a number* – не число).

С точки зрения интерпретации массивов в двумерном представлении (как набор строк и столбцов) можно считать, что приведенные выше массивы содержат в себе по 3 "слоя" (первый индекс в размерности), каждый из которых представляет собой 4 строки (второй индекс) и 5 столбцов (третий). Каждый из "слоев" при выводе массива на печать оказывается автоматически отделен от следующего пустой строкой – для удобства чтения.

### Обращение к элементам массива

Итак, мы создали ряд массивов. Для того, чтобы обратиться к какому-либо элементу массива, используется тот же способ, что и в случае последовательностей: после указания имени массива необходимо в квадратных скобках ввести индексы нужного элемента (отсчитываемые от нуля, как и прежде).

Индексы указываются внутри одной пары скобкок через запятую:

~~~python
a[2, 1, 0]
~~~

– в отличие от многомерной последовательности данных, задаваемой, например, с помощью вложенных списков. Для сравнения: если бы массив `a` был реализован именно как список, обращение к его элементам происходило бы так:

~~~python
a_list[2][1][0]
~~~

Для хранения и создания подобного объекта потребовалось бы использование трёхкратно (!) вложенных списков, что гораздо сложнее создания массива как с точки зрения реализации кода, так и с вычислительной позиции (упомянутая выше скорость обработки данных). При работе с многомерными данными, размерность которых превышает 3, сложность подобной задачи стремительно увеличивается!

Вернёмся к массивам `NumPy`. Поскольку тип `ndarray` – **изменямый** тип данных, то, обращаясь к элементам массива, можно не только получать информацию о них, но и менять значения, которые в них хранятся, например:

~~~python
a[2, 1, 0] = 5
~~~

##### Срезы
Ещё одной важной особенностью синтаксиса массивов `NumPy` является возможность использования срезов при работе с индексами массива.<br>
Сравните:

In [2]:
# работа со срезами
import numpy as np

# создадим новый пробный массив "a"
a = np.reshape(np.linspace(1, 12, 12), (3, 4))

# выведем теперь на печать как массив "a", так и комбинации его элементов
print('a:', a, sep='\n', end='\n\n')

print('элемент (0, 0):', a[0, 0])
print('элемент (2, 2):', a[2, 2])
print('строка 1: ', a[1, :])
print('столбец 2:', a[:, 2])

print('\nокно 2x2 (строки 1–2 и столбцы 0–1):', a[1:3, 0:2], sep='\n')


a:
[[ 1.  2.  3.  4.]
 [ 5.  6.  7.  8.]
 [ 9. 10. 11. 12.]]

элемент (0, 0): 1.0
элемент (2, 2): 11.0
строка 1:  [5. 6. 7. 8.]
столбец 2: [ 3.  7. 11.]

окно 2x2 (строки 1–2 и столбцы 0–1):
[[ 5.  6.]
 [ 9. 10.]]


Разберем некоторые составляющие этого кода:

Команда `np.reshape()` (буквально, "поменяй форму") нужна для преобразования размерности получившегося массива `a`:
* первый аргумент функции – массив, размерность которого преобразуется.
* второй аргумент функции – кортеж, указывающий размерность нового массива, в который мы преобразуем исходный.

Команда `np.linspace(<начало>, <конец>, <количество>)` – это одномерная цепочка заданного *<количества>* равноотстоящих друг от друга вещественных чисел на промежутке от *<начала>* до *<конца>*.

В примере выше **преобразуемый** массив (первый аргумент) – это цепочка чисел от 1.0 до 12.0 *включительно* (ответьте самостоятельно, почему). Функция `np.reshape()` преобразует данную цепочку размерности (1, 12) – "1 строка и 12 столбцов" – в массив размерности (3, 4) – "3 строки и 4 столбца". Данный процесс происходит последовательно: первые 4 элемента цепочки попадают в первую строку, затем вторые 4 элемента – во вторую, и заключительные – в третью. Поскольку такое распределение арифметически возможно, ошибки не возникает.

Зато представленный ниже вариант не будет успешно выполнен. В массивах `NumPy` не может быть незаполненных ячеек в массиве заданного размера:

In [2]:
np.reshape(np.linspace(1, 11, 11), (3, 4))


ValueError: cannot reshape array of size 11 into shape (3,4)

А теперь продемонстрируем более сложный функционал срезов на примере массива `b`:

In [6]:
b = np.ndarray((3, 4, 5), dtype = float)

# добавим эти элементы как маркеры – чтобы можно было проверить себя
b[::-1, ::2, 2:3] = 12345
b[1:3, 0, 2] = 777

print('массив b')
print(b)
print('\n' + '-'*70 + '\n')
print('срез по первой оси', b[1:3, 0, 2], sep='\n', end='\n\n')
print('срез по трём осям', b[::-1, ::2, 2:3], sep='\n', end='\n\n')
print('срезы по первой и третьей осям', b[:, 2, :], sep='\n')

b.shape


массив b
[[[1.05281876e-311 1.05281713e-311 1.23450000e+004 0.00000000e+000
   8.44852254e-322]
  [2.10966031e-321 9.06092204e-312 2.49260694e-306 4.31813374e-321
   0.00000000e+000]
  [7.04446241e-191 0.00000000e+000 1.23450000e+004 1.44932313e-311
   3.64993935e-157]
  [0.00000000e+000 5.41946267e-291 1.62017374e-248 2.84453558e-310
   1.76083582e-310]]

 [[0.00000000e+000 0.00000000e+000 7.77000000e+002 0.00000000e+000
   0.00000000e+000]
  [1.38274488e-316 2.41907523e-311 1.87544878e-128 0.00000000e+000
   0.00000000e+000]
  [9.06129374e-312 6.13068412e-046 1.23450000e+004 0.00000000e+000
   0.00000000e+000]
  [1.44935952e-311 1.02711249e-243 3.69148781e-316 2.14109375e-310
   0.00000000e+000]]

 [[0.00000000e+000 1.38275753e-316 7.77000000e+002 2.27239638e-315
   0.00000000e+000]
  [2.49208272e-306 3.72031431e-320 0.00000000e+000 3.62861280e-312
   9.06092203e-312]
  [3.15399197e-234 0.00000000e+000 1.23450000e+004 2.10966031e-321
   3.43607053e-315]
  [3.62861280e-312 4.10453931e

(3, 4, 5)

Разберем, что мы получили.

1. `b[1:3, 0, 2]` звучит буквально так: из массива `b` возьми слои с первого (не нулевого) по третий не включительно, затем в каждом из них выбери нулевую строчку и второй столбец. Посмотрите на массив `b` и убедитесь, что результат верный.
2. `b[:, 2, :]` означает: возьми все слои, и в каждом выбери вторую строку и все столбцы.
3. Чуть более сложный вариант: `b[::-1, ::2, 2:3]` предлагаем вам озвучить самостоятельно. Вспомните, как работают срезы и что означает второе двоеточие.

Теперь давайте рассмотрим примеры обращения к элементам массива и их модификации (в т. ч. с использованием срезов).

In [6]:
import numpy as np

a = np.ndarray((2, 5), dtype=float)      # создаём новый пробный массив
print('Исходный массив a', a, sep='\n')  # выведем его на печать

# занулим все элементы массива, стоящие на первой строке,
# а на второй – разместим единицы
a[0, :] = 0
a[1, :] = 1
print('\nЗанулили первую строку, по второй строке – единицы', a, sep='\n')

# сделаем каждый второй элемент первой строки равным 2
a[0, 1::2] = 2
print('\nКаждый второй элемент первой строки равен 2', a, sep='\n')

# перемешаем элементы строк массива в шахматном порядке
# без добавления ".copy()" результат будет не совсем таким, как мы ожидаем!!!
a[0, ::2], a[1, ::2] = a[1, ::2].copy(), a[0, ::2].copy()
print('\nПеремешаем элементы массива', a, sep='\n')

# используем выражение-генератор для изменения элементов 1 и 2 строк массива
a[0, 1::2] = [i+5 for i in range(len(a[0, :]))][1::2]
a[1, :]    = [i+1 for i in range(len(a[1, :]))]
print('\nЭлементы массива заданы выражениями-генераторами', a, sep='\n')


Исходный массив a
[[0.00000000e+000 0.00000000e+000 6.79038653e-313 6.79038653e-313
  6.79038653e-313]
 [1.93101617e-312 9.33678148e-313 1.14587773e-312 1.97345609e-312
  2.02566915e-322]]

Занулили первую строку, по второй строке – единицы
[[0. 0. 0. 0. 0.]
 [1. 1. 1. 1. 1.]]

Каждый второй элемент первой строки равен 2
[[0. 2. 0. 2. 0.]
 [1. 1. 1. 1. 1.]]

Перемешаем элементы массива
[[1. 2. 1. 2. 1.]
 [0. 1. 0. 1. 0.]]

Элементы массива заданы выражениями-генераторами
[[1. 6. 1. 8. 1.]
 [1. 2. 3. 4. 5.]]


### Специальные массивы:

Наряду с массивами общего типа, в модуле `NumPy` определены массивы специального типа. К ним относятся:
1. Нулевая матрица: `zeros(sh, dt)`;
2. Единичная матрица: `ones(sh, dt)`;
3. Единично-диагональная (1 на главной диагонали, 0 для других элементов):
    - `еуе(n, m)` для матрицы $n \times m$
    - `еуе(n)` или `identity(n)` – для квадратной матрицы.

В этих функциях:
* `sh` – **кортеж**, задающий размерность массива (shape),
* `dt` – параметр, определяющий тип данных (data type),
* `n`, `m` – некоторые натуральные числа.

Попробуем воспользоваться этими функциями.

In [3]:
import numpy as np

a0 = np.zeros((2, 3), 'float')
print('a0:', a0, sep='\n')

a1 = np.ones((5, 8), 'int')
print('\na1:', a1, sep='\n')

a2 = np.eye(4, 6, dtype = int)
print('\na2:', a2, sep='\n')

a3 = np.identity(4, 'float')
print('\na3:', a3, sep='\n')


a0:
[[0. 0. 0.]
 [0. 0. 0.]]

a1:
[[1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]]

a2:
[[1 0 0 0 0 0]
 [0 1 0 0 0 0]
 [0 0 1 0 0 0]
 [0 0 0 1 0 0]]

a3:
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


### meshgrid

К ним же отнесём способ создания таблиц перекрёстных значений `np.meshgrid(values, indexing='xy')`. Она ожидает перечисление возможных значений каждой величины и способ комбинирования('xy' или 'ij'). На выходе таблица со всеми комбинациями переданных значений массивов - создаёт пространственную сетку

In [2]:
import numpy as np

x = np.linspace(1, 4, 4)
y = np.linspace(5, 8, 4)
print(x, y)

np.meshgrid(x, y)

[1. 2. 3. 4.] [5. 6. 7. 8.]


(array([[1., 2., 3., 4.],
        [1., 2., 3., 4.],
        [1., 2., 3., 4.],
        [1., 2., 3., 4.]]),
 array([[5., 5., 5., 5.],
        [6., 6., 6., 6.],
        [7., 7., 7., 7.],
        [8., 8., 8., 8.]]))

В первом массиве чередуются все элементы `x`, тогда как во втором поочерёдно каждый из элементов `y` повторояется `len(x)` раз. Теперь разберёмся с разницей в индексации:

In [27]:
x = np.linspace(1, 2, 2)
y = np.linspace(3, 4, 2)
z = np.linspace(5, 6, 2)

a = np.meshgrid(x, y, z, indexing='xy')
b = np.meshgrid(x, y, z, indexing='ij')

print(a, b, sep='\n\n')

(array([[[1., 1.],
        [2., 2.]],

       [[1., 1.],
        [2., 2.]]]), array([[[3., 3.],
        [3., 3.]],

       [[4., 4.],
        [4., 4.]]]), array([[[5., 6.],
        [5., 6.]],

       [[5., 6.],
        [5., 6.]]]))

(array([[[1., 1.],
        [1., 1.]],

       [[2., 2.],
        [2., 2.]]]), array([[[3., 3.],
        [4., 4.]],

       [[3., 3.],
        [4., 4.]]]), array([[[5., 6.],
        [5., 6.]],

       [[5., 6.],
        [5., 6.]]]))


При формате `'ij'` наоборот точки заполняются с изменения последних осей, можем продемонстрировать для 3-х мерного случая в виде рисунка:

### Математические функции

До сих пор мы рассматривали элементарные операции создания массива и обращения к его элементам. Наряду с этими операциями в модуле `NumPy` реализованы и всевозможные математические функции для работы с элементами массивов (избавляющие пользователя от необходимости последовательного обращения к каждому элементу массива для обработки содержащихся в них данных). 

Краткий (и отнюдь не исчерпывающий) список таких функций представлен ниже. Пусть `x` и `y` – некоторые `NumPy`-массивы. Тогда к ним применимы следующие действия:

|Функция |Описание: поэлементное... |Функция |Описание: возвращает поэлементно...|
|---|:-|:-|:-|
|`np.add(x, у)`|сложение двух массивов|`np.angle(x)`|углы комплексных чисел|
|`np.subtract(х, у)`|вычитание|`np.phase(x)`|фазы комплексных чисел|
|`np.multiply(x, у)`|умножение|`np.real(x)`|реальные части|
|`np.divide(x, у)`|деление|`np.imag(x)`|мнимые части|
|`np.power(x, у)`|возведение в степень|`np.conj(x)`|комплексно сопряженные|
|`np.negative(x, у)`|изменение знака в массиве `x` и запись в массив `y`|

Рассмотрим примеры:

In [8]:
import numpy as np
import math

# единичный массив
a = np.ones((3, 4))

# создадим пробный массив, содержащий элементы ~pi
b = np.array([(i+1)*j*math.pi/12 for i in range(3) for j in range(4)])
# и переформатируем его под размерность "a"
b = np.reshape(b, (3, 4))
print('a =', a, sep='\n')
print('\nb =', b, sep='\n')

# сумма массивов a и b
c = np.add(a, b)
print('\na + b =', c, sep='\n')

# сумму можно задать и так
c = a + b
print('\na + b =', c, sep='\n')

# поэлементное возведение элементов массива "c"
# в степени из элементов массива "b"
d = np.power(c, b)
print('\n(a + b)^b =', d, sep='\n')

# массив, элементы которого есть sin(b_ij)
e = np.sin(b)
print('\nsin(b_ij) =', e, sep='\n')


a =
[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]

b =
[[0.         0.26179939 0.52359878 0.78539816]
 [0.         0.52359878 1.04719755 1.57079633]
 [0.         0.78539816 1.57079633 2.35619449]]

a + b =
[[1.         1.26179939 1.52359878 1.78539816]
 [1.         1.52359878 2.04719755 2.57079633]
 [1.         1.78539816 2.57079633 3.35619449]]

a + b =
[[1.         1.26179939 1.52359878 1.78539816]
 [1.         1.52359878 2.04719755 2.57079633]
 [1.         1.78539816 2.57079633 3.35619449]]

(a + b)^b =
[[ 1.          1.06276979  1.24666805  1.57656642]
 [ 1.          1.24666805  2.11760879  4.40689589]
 [ 1.          1.57656642  4.40689589 17.33796361]]

sin(b_ij) =
[[0.         0.25881905 0.5        0.70710678]
 [0.         0.5        0.8660254  1.        ]
 [0.         0.70710678 1.         0.70710678]]


Обратите внимание на использование двойного генератора списка (list comprehension) в этом случае:
~~~python
b = np.array([(i+1)*j*math.pi/12 for i in range(3) for j in range(4)])
~~~

создает 3 строки (первый генератор, индекс `i`), в каждую из которых записывает 4 элемента (второй генератор, индекс `j`), определяемые текущими величинами `i` и `j`.

## Несколько заключительных слов

Перечисленные методы, конечно же, не исчерпывают всего множества возможностей, которые предоставляет в распоряжение исследователя модуль `NumPy`. Для полного списка функционала модуля рекомендуем обратиться к документации библиотеки, расположенной по [этому адресу](https://numpy.org/doc/1.23/user/index.html) (на момент подготовки материала лекции наиболее свежая версия `NumPy` – 2.1).